client/web/src/hooks/self-update.ts

```diff
@@ -0,0 +1,114 @@
+import { useEffect, useState } from "react"
+import { apiFetch } from "src/api"
+
+// see tailcfg.ClientVersion
+export type ClientVersion = {
+  RunningLatest: boolean
+  LatestVersion?: string
+}
+
+// see ipnstate.UpdateProgress
+export type UpdateProgress = {
+  status: "UpdateFinished" | "UpdateInProgress" | "UpdateFailed"
+  message: string
+  version: string
+}
+
+export enum UpdateState {
+  UpToDate,
+  Available,
+  InProgress,
+  Complete,
+  Failed,
+}
+
+export function useInstallUpdate(currentVersion: string, cv: ClientVersion) {
+  const [updateState, setUpdateState] = useState<UpdateState>(
+    cv.RunningLatest ? UpdateState.UpToDate : UpdateState.Available
+  )
+
+  const [updateLog, setUpdateLog] = useState<string>("")
+
+  const appendUpdateLog = (msg: string) => {
+    setUpdateLog(updateLog + msg + "\n")
+  }
+
+  useEffect(() => {
+    if (updateState !== UpdateState.Available) {
+      // useEffect cleanup function
+      return () => {}
+    }
+    apiFetch("/update", "POST").catch((err) => {
+      console.log(err)
+      setUpdateState(UpdateState.Failed)
+    })
+
+    setUpdateState(UpdateState.InProgress)
+
+    let tsAwayForPolls = 0
+    let updateMessagesRead = 0
+
+    let timer = 0
+
+    function poll() {
+      apiFetch("/update/progress", "GET")
+        .then((res) => res.json())
+        .then((res: UpdateProgress[]) => {
+          for (; updateMessagesRead < res.length; ++updateMessagesRead) {
```

Comment:
could do `res.forEach` instead

Reply:
not using `forEach` here because `updateMessagesRead` is stored across pings to that endpoint. I can add a comment explaining what's happening in more detail